### PR TITLE
EEG-GCNN: Dataset, Task, Models, and Neurological Disease Detection Pipeline

### PR DESCRIPTION
**Pull request by jburhan**
## Contributors
- Jimmy Burhan — jburhan2 (Dataset + Task)
- Robert Coffey — rc37 (Models + Training)

## Type of Contribution
**Option 4: Full Pipeline** — Dataset + Task + Model (60 pts)

## Paper
[Wagh & Varatharajah (2020) — EEG-GCNN: Augmenting Electroencephalogram-based Neurological Disease Diagnosis using a Domain-guided Graph Convolutional Neural Network, ML4H @ NeurIPS 2020](https://proceedings.mlr.press/v136/wagh20a.html)

## Summary

End-to-end EEG-GCNN pipeline for neurological disease detection from EEG recordings, replicating and extending Wagh & Varatharajah (2020).

### Dataset & Task (Jimmy Burhan — jburhan2)
- `EEGGCNNRawDataset` — raw EEG loader for TUAB (1,385 clinically-normal patients, label 0) + LEMON (208 healthy volunteers, label 1): resamples to 250 Hz, 1 Hz high-pass + 60/50 Hz notch filter, 8 bipolar channels, 10 s non-overlapping windows, 6-band Welch PSD features (delta/theta/alpha/beta/low_gamma/high_gamma), writes 5 FigShare-format arrays via `precompute_features()`
- `EEGGCNNDataset` — FigShare pre-computed dataset (1,593 subjects, 225,334 windows); builds graph adjacency via tunable blend of geodesic distance and spectral coherence: `edge_weight = α·geodesic + (1−α)·coherence`
- `EEGGCNNDiseaseDetection` — per-sample streaming task with configurable adjacency type (spatial / functional / combined) and band subset; available as an alternative on-the-fly entry point
- `EEGGCNNClassification` — window-level classification task for FigShare data; supports `excluded_bands` for spectral ablation

### Models & Training (Robert Coffey — rc37)
- `EEGGraphConvNet` — two-layer GCN classifier (reproduces paper's Shallow EEG-GCNN, float32)
- `EEGGATConvNet` — two-layer graph attention network classifier (float64)
- Training pipelines, holdout evaluation scripts, band frequency ablation script
- **83 unit and integration tests** across 4 test files — synthetic fixtures for fast unit coverage + real EDF/BrainVision integration tests via `TestRawDatasetPrecompute`

### Ablation 1 — Adjacency Type (α sweep, GCN)
Varied α ∈ {0.0, 0.25, 0.5, 0.75, 1.0} on the full FigShare dataset (477 test patients, 70/30 split, 10-fold CV). Paper only uses α=0.5 — this is our novel contribution:

| α | auroc_patient | bal_acc | f1 |
|---|---|---|---|
| 0.0 (coherence only) | 0.8841 ± 0.0132 | 0.8087 ± 0.0130 | 0.8222 ± 0.0176 |
| 0.25 | 0.8943 ± 0.0072 | 0.8069 ± 0.0122 | 0.8385 ± 0.0162 |
| **0.50 (paper default)** | **0.8984 ± 0.0102** | **0.8276 ± 0.0122** | 0.8399 ± 0.0281 |
| **0.75** | **0.8991 ± 0.0066** | 0.8209 ± 0.0051 | **0.8491 ± 0.0323** |
| 1.0 (geodesic only) | 0.7891 ± 0.0471 | 0.7406 ± 0.0349 | 0.8211 ± 0.0599 |

α=1.0 (spatial-only) collapses performance — functional connectivity is essential. α=0.75 best for auroc_patient and F1; α=0.50 best for balanced accuracy.

### Ablation 2 — Spectral Frequency Analysis (GCN, inference-only)
Leave-one-out (LOO) and keep-one-in (KOI) across all 6 bands using trained GCN checkpoints — no retraining required. Low gamma is the single most important band (removing it drops AUROC from 0.898 → 0.594; keeping only it recovers 0.723). All bands make a meaningful contribution — none is redundant. Numerical results in project slides.

### Best Results (GAT, tuned learning rate)
| auroc_patient | bal_acc | f1 | precision | recall |
|---|---|---|---|---|
| 0.9415 ± 0.0251 | 0.8854 ± 0.0322 | 0.9201 ± 0.0231 | 0.9855 ± 0.0089 | 0.8638 ± 0.0409 |

GAT with tuned LR outperforms GCN on every patient-level metric.

## File Guide

| File | Description |
|------|-------------|
| `pyhealth/datasets/eeg_gcnn.py` | `EEGGCNNDataset` — FigShare pre-computed loader |
| `pyhealth/datasets/eeg_gcnn_raw.py` | `EEGGCNNRawDataset` — raw TUAB + LEMON loader with `precompute_features()` |
| `pyhealth/datasets/configs/eeg_gcnn.yaml` | FigShare dataset config |
| `pyhealth/datasets/configs/eeg_gcnn_raw.yaml` | Raw dataset config |
| `pyhealth/tasks/eeg_gcnn_classification.py` | `EEGGCNNClassification` — window-level task for FigShare path |
| `pyhealth/tasks/eeg_gcnn_disease_detection.py` | `EEGGCNNDiseaseDetection` — streaming task for raw path |
| `pyhealth/models/eeg_gcnn.py` | `EEGGraphConvNet` (GCN) |
| `pyhealth/models/eeg_gatcnn.py` | `EEGGATConvNet` (GAT) |
| `examples/eeg_gcnn/eeg_gcnn_classification_gcn_precompute.py` | Precompute: raw TUAB + LEMON → 5 FigShare-format arrays (GCN path) |
| `examples/eeg_gcnn/eeg_gcnn_classification_gcn_training.py` | GCN training (10-fold CV) |
| `examples/eeg_gcnn/eeg_gcnn_classification_gcn_evaluation.py` | GCN holdout evaluation (Youden's J threshold) |
| `examples/eeg_gcnn/eeg_gcnn_classification_gcn_band_ablation.py` | Spectral frequency ablation (LOO + KOI, 13 conditions) |
| `examples/eeg_gatcnn/eeg_gcnn_classification_gat_precompute.py` | Precompute for GAT path (float64) |
| `examples/eeg_gatcnn/eeg_gcnn_classification_gat_training.py` | GAT training (tuned LR) |
| `examples/eeg_gatcnn/eeg_gcnn_classification_gat_evaluation.py` | GAT holdout evaluation |
| `examples/eeg_gcnn/sample_precomputed_data/` | 5 FigShare-format sample arrays for quick-start testing |
| `examples/eeg_gcnn/sample_raw_data/` | Sample EDF (TUAB) + BrainVision (LEMON) files used by integration tests |
| `tests/core/test_eeg_gcnn_raw_dataset.py` | 45 tests: `EEGGCNNRawDataset` + `EEGGCNNDiseaseDetection` (synthetic + real EDF/BrainVision integration) |
| `tests/core/test_eeg_gcnn_dataset.py` | 18 tests: `EEGGCNNDataset` + `EEGGCNNClassification` (synthetic fixtures) |
| `tests/core/test_eeg_gcnn_model.py` | 8 tests: `EEGGraphConvNet` (forward pass, shapes, gradients) |
| `tests/core/test_eeg_gat_model.py` | 12 tests: `EEGGATConvNet` (forward pass, dropout configs, gradients) |
| `docs/api/eeg_gcnn_pipeline.rst` | Full pipeline documentation |
| `docs/api/datasets/pyhealth.datasets.EEGGCNNDataset.rst` | API doc — FigShare dataset |
| `docs/api/datasets/pyhealth.datasets.EEGGCNNRawDataset.rst` | API doc — raw dataset |
| `docs/api/models/pyhealth.models.EEGGraphConvNet.rst` | API doc — GCN model |
| `docs/api/models/pyhealth.models.EEGGATConvNet.rst` | API doc — GAT model |
| `docs/api/tasks/pyhealth.tasks.EEGGCNNClassification.rst` | API doc — classification task |
| `docs/api/tasks/pyhealth.tasks.eeg_gcnn_disease_detection.rst` | API doc — disease detection task |

## Test Plan
```bash
# Raw path — EEGGCNNRawDataset + EEGGCNNDiseaseDetection (45 tests)
pytest tests/core/test_eeg_gcnn_raw_dataset.py -v

# FigShare path — EEGGCNNDataset + EEGGCNNClassification (18 tests)
pytest tests/core/test_eeg_gcnn_dataset.py -v

# Models — EEGGraphConvNet + EEGGATConvNet (20 tests)
pytest tests/core/test_eeg_gcnn_model.py tests/core/test_eeg_gat_model.py -v

# All 83 tests, ~30 seconds, no real EEG data required for synthetic tests
pytest tests/core/test_eeg_gcnn_raw_dataset.py tests/core/test_eeg_gcnn_dataset.py tests/core/test_eeg_gcnn_model.py tests/core/test_eeg_gat_model.py -v
```

### CODE REVIEW

**Comment by jburhan** [comment]
Closing to reopen from the correctly named branch jburhan:jburhan-rcoffey-contribution.
